### ISSUE TITLE
curl can't download from tlsv1.2 only site

### ISSUE THREAD

**Issue by mikkeloscar** [travis-ci/travis-ci#4757]
Trying to download a file with curl over https from a server only supporting tlsv1.2 results in this error:

```
* About to connect() to arch.localmsp.org port 443 (#0)
*   Trying 68.168.184.57...   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0connected
* successfully set certificate verify locations:
*   CAfile: none
  CApath: /etc/ssl/certs
* SSLv3, TLS handshake, Client hello (1):
} [data not shown]
* SSLv3, TLS alert, Server hello (2):
{ [data not shown]
* error:14077410:SSL routines:SSL23_GET_SERVER_HELLO:sslv3 alert handshake failure
* Closing connection #0
curl: (35) error:14077410:SSL routines:SSL23_GET_SERVER_HELLO:sslv3 alert handshake failure
```
https://travis-ci.org/mikkeloscar/arch-travis/builds/78991099

The file in this example is here: https://arch.localmsp.org/arch/iso/2015.09.01/archlinux-bootstrap-2015.09.01-x86_64.tar.gz

The server only supports tlsv1.2: https://www.ssllabs.com/ssltest/analyze.html?d=arch.localmsp.org

**Comment by BanzaiMan** [comment]
Looks like the binary on Precise image doesn't have TLS configured. The recent docker image runs on Trusty, and it supports TLS v1.

**Comment by mikkeloscar** [comment]
It does work for servers supporting TLSv1.0 and TLSv1.1 like this one: https://www.ssllabs.com/ssltest/analyze.html?d=ftp.lysator.liu.se&s=130.236.254.50

I can use that as a work around since I need `sudo` and thus can't use the container based environment yet.

**Comment by kofalt** [comment]
Did anyone figure out a workaround for this? 

I ran into this also, and discovered it is not possible to download MongoDB assets in the container infra.

**Comment by jayvdb** [comment]
The primary problem is that the precise image is from Feb 2015, and the support for tls 1.2 wasnt re-added until ~May 2015 : http://www.ubuntu.com/usn/usn-2606-1/

I went looking for solutions, and built this testing environment to see what fails for various configurations accessing `https://proofwiki.org/wiki/Main_Page`, which only supports TLS 1.2 (see https://www.ssllabs.com/ssltest/analyze.html?d=www.proofwiki.org)

Results at https://travis-ci.org/jayvdb/my-ci-test/builds/131596507 ; cliff notes:

On precise containers, out of the box, the following works

```python
import ssl, urllib3
conn = urllib3.connection_from_url(
    'https://proofwiki.org/wiki/Main_Page',
    ssl_version=ssl.PROTOCOL_TLSv1_2,
)
r = conn.request('GET', 'https://proofwiki.org/wiki/Main_Page')
print(r.status)
```
The same can be achieved using `requests`, but it requires [more code](http://docs.python-requests.org/en/master/user/advanced/#example-specific-ssl-version).

Command line tools like `wget` and `curl` fail, and most libraries in all languages will also fail if the protocol isnt explicitly requested, unless the library has specific support for attempting to connect using TLSv1.2 without it being specified like the urlllib3 example above (e.g. it downgrades when a normal automatic negotiation fails), but it would be generally bad for a library to do that.

On precise with sudo enabled, there are a few more options .e.g.
```bash
python -c "import requests; print(requests.get('$URL').status_code);"
wget $URL
curl $URL
```

However on precise with sudo enabled, explicitly requesting a protocol from `wget` & `curl` fail, as the necessary command line option didnt exist yet.

On trusty, everything works wonderfully except for `wget`, which fails no matter what I try.  So use `curl` if you want a command line fetch tool.

**Comment by jayvdb** [comment]
Minor update, the better way to do this on Python is with an isolated `PoolManager`, and not using `connection_from_url`:

```python
import ssl, urllib3
pm = urllib3.PoolManager(ssl_version=ssl.PROTOCOL_TLSv1_2)
r = pm.request('GET', 'http://proofwiki.org/wiki/Main_Page')
print(r.status)
```

And I ran into trouble doing the same with `requests` (https://github.com/kennethreitz/requests/issues/3214 filed) 

Note that only works on Python 2.7.9+ and Python 3.4+ as `ssl.PROTOCOL_TLSv1_2` is missing on lower versions of Python 2 and Python 3.

Latest demo at https://travis-ci.org/jayvdb/my-ci-test/builds/132022037

**Comment by jayvdb** [comment]
The magic incantation to workaround this on precise containers is:

```yaml
addons:
  apt:
    sources:
      - debian-sid
    packages:
      - libssl1.0.0
```

**Comment by Godin** [comment]
@jayvdb unfortunately this workaround leads to failure when installing `g++-multilib`:
```yml
addons:
  apt:
    sources:
      - debian-sid
    packages:
      - libssl1.0.0
      - g++-multilib
```

**Comment by jayvdb** [comment]
Yup; that happens a lot when using `debian-sid`.  It should also be possible to compile `libssl` from source, and ensure the new built library is used by way of `LD_LIBRARY_PATH`.

**Comment by BanzaiMan** [comment]
The current work around is to move off Precise containers. https://github.com/travis-ci/travis-ci/issues/7131#issuecomment-272240670

**Comment by marcelstoer** [comment]
Since I wasn't aware of this particular issue I contacted support and was told

> However, our Engineering Team confirmed that this will be fixed with the next update to our Precise containers for which we don't have an ETA at this time. We will update the issue when the new Precise containers are released.

**Comment by BanzaiMan** [comment]
This may be fixed by updating `libssl1.0.0`:

```yaml
addons:
  apt:
    packages:
      - libssl1.0.0
```

We may add this workaround for Precise container builds in the near future.
